### PR TITLE
Algorithm for spreading staves/systems over a page

### PR DESCRIPTION
**Pull request by njvdberg**
This PR implements [automatic vertical justification of staves](https://trello.com/c/feV4aI2I/6-implement-automatic-vertical-justification-of-staves) of the [MuseScore Engraving Push](https://trello.com/b/JVGhCNm6/musescore-engraving-push)

The algorithm
-------------
After the normal layout of a page, a new algorithm tries the spread the staves in such way the space between the lowest staff and the bottom margin of the page is used too by adding extra space between the staves.
Not all gaps between the staves will get the same amount of space but the algorithm distinguishes 3 types of gaps:

1) Gaps between the systems.

2) Gaps between the staves within a system, grouped by a [ bracket.

3) Gaps between the staves witinh a system, but outside a [ bracket.

The space between the lowest staff and the bottom margin of the page is divided over these gaps using a weight factor for each of these types of gaps.

For an example of how this algorithm works, see these 2 pages of a score:

before

This pages contains 12 staves in 2 systems.
The applied weight factors were the defaults and the number of gaps to which the factors apply:

 1. Between systems: 6<br>1 gap (between the systems).
 2. Within systems, grouped by a [ bracket: 1.5<br>2 * 1 gap:
  between <code>Vln.</code> and <code>Vc.</code>
 3. Within systems, not grouped by a [ bracket: 3<br>2 * 4 gaps
  between <code>A. Sax.</code> and <code>Fl.</code>
  between <code>Fl.</code> and <code>Vln.</code>
  between <code>Vc.</code> and <code>Pno</code>
  between the two staves of the <code>Pno.</code>

Please note that the gap above the first staff of the system is not taken into account, this space remains unchanged.
At the bottom of the page, there is quite some space left between the lowest staff of the second system and the bottom music margin. The algorithm is designed to reduce this empty space by adding extra space between the gaps:
<code>"space left" * "weight factor for gap" / "sum of all weight factors on page"</code>

So, the between each <code>A. Sax.</code> and <code>Fl.</code>, <code>3/33 * "space left"</code>, so roughly 9% of the space left is added. Between the systems, 18% is added.

While adding this extra space, the maximum distance is obeyed. As a result, the algorithm can run in multiple passes. The algorithm stops:

 - When there is no space left,
 - All gaps reached their maximum distance or
 - After 10 passes (preventing an endless loop).

When calculating the <code>"sum of all weight factors on page"</code>, gaps which are already at their maximum space are ignored.

The passes won't actually changes anything, just an offset for each staff is calculated. When the algorithm finished the extra distance is applied to all staves.

The end result will looks like:

after

New style properties
--------------------
To control the algorithm 5 new style properties are introduced:

 1. <code>spreadFactorSystem</code> (default: 6.0)<br>This parameter specifies the weight factor for the space between system.<br>On the form this is the property <code>Factor distance between systems</code>

 2. <code>spreadFactorSpacing</code> (default: 3.0)<br>This parameter specifies the weight factor for the space between staves within a system. On the form this is the property <code>Factor distance within system</code>

 3. <code>spreadFactorGroup</code> (default: 1.5)</br>This parameter specifies the weight factor for the space between staves within a system *and* grouped by a [  bracket.<br>On the form this is the property <code>Factor distance staves grouped by [ bracket</code>

 4. <code>maxStaffDistance</code> (default: 9.5)</br>This parameter specifies the maximum staff distance between staved within a system.<br>On the form this is the property <code>Max. staff distance</code>

 5. <code>maxAkkoladeDistance</code> (default: 6.5)</br>This parameter specifies the maximum distance between staves belonging to the same instrument. Usually these staves are group by a brace (<code>{</code>) but also when staves are added to an instrument, this parameter defines the maximum distance between the staves of the instrument.<br>On the form this is the property<code>Max. grand staff distance:</code>

All these parameters are on the <code>Page</code> page of the <code>Style</code> form.

The algorithm is disabled if all 3 spread factors, <code>spreadFactorSystem</code>, <code>spreadFactorSpacing</code> and <code>spreadFactorGroup</code> are 0.0.

Compatibility
-------------
When loading a score, there is no way to tell whether this is an old score without using the new spreading algorithm, or a new score using the spreading algorithm using the default properties.
To prevent the spreading algorithm is activated on old scores, a new tag <code>verticalAdjustStaves</code> is written in new scores. The absens of this tag indicates the score is an old score and during load, all spread factors (<code>spreadFactorSystem</code>, <code>spreadFactorSpacing</code> and <code>spreadFactorGroup</code>) are set to 0.0, hence disabling the algorithm.

This is also the reason why this PR is that big, all references had to modified.

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

**Comment by IsaacWeiss** [comment]
I love this! Due to the changes to the ref files, I would appreciate it if #6365 could be merged first, so as not to necessitate more conflict resolutions on that branch (indeed, I see that there is already a new conflict that I will need to address).

File: libmscore/layout.h [hunk 1]
```diff
@@ -13,11 +13,61 @@
...
 class Segment;
 class Page;
 
+//---------------------------------------------------------
+//   VerticalStretchData
+//    helper class for spreading staves over a page
+//---------------------------------------------------------
+
+class VerticalGapData {
+      char      c1             { ' '     };
```
Comment (Marr11317) [inline libmscore/layout.h:29]: I see these are set, but not when they are used. Could you clarify their usage with a comment?

Comment (Marr11317) [reply libmscore/layout.h:29]: I'm speaking about `c1`, `c2`, etc. here

File: libmscore/score.cpp [hunk 2]
```diff
@@ -3430,6 +3430,31 @@ void Score::selectSimilarInRange(Element* e)
...
 
+//---------------------------------------------------------
+//   isStaffSpreadEnable
+//---------------------------------------------------------
+
+bool Score::isStaffSpreadEnable() const
+      {
+      // Don't use equal tests on floating numbers.
+      // (x < zeroTreshold) is considered to be 0.0.
+      const qreal zeroTreshold { 0.001 };
```
Comment (Marr11317) [inline libmscore/score.cpp:3441]: why use an initializer list for a single qreal?

**Comment by Jojo-Schmitz** [comment]
> I love this! Due to the changes to the ref files, I would appreciate it if #6365 could be merged first, so as not to necessitate more conflict resolutions on that branch (indeed, I see that there is already a new conflict that I will need to address).

PR #6365 is for the master branch, this here for the 3.x branch though

File: libmscore/layout.h (hunk 1 above)
Comment (njvdberg) [reply libmscore/layout.h:29]: Sorry, leftovers of debugging code, will be removed.

File: libmscore/score.cpp (hunk 2 above)
Comment (njvdberg) [reply libmscore/score.cpp:3441]: Of the three forms of initialization (copy, direct and brace) the used (direct) brace initialization is the recommended form (according the [C++ course](https://www.learncpp.com/cpp-tutorial/variable-assignment-and-initialization/) I did some time ago). I use it more an more nowadays.
So, in fact it isn't an initializer list but a brace initialization (which also can be a list, depending on the type of variable you are initializing).

**Comment by njvdberg** [comment]
> (I assume it is intended for MuseScore 3.6? As it does change translatable string it certainly is not for 3.5.1.)

Yes, it is for 3.6.

**Comment by oktophonie** [comment]
This is fantastic!
I'd propose some changes to the Style->Page settings, something like:

'Staff distance' should be named 'Min. staff distance', and the default should be a small value, like 1.5.
'Grand staff distance' becomes 'Min. grand staff distance', with same default, 1.5.
'Min. system distance' could have a lower default too, like 6. These low values don't just mean that closer spacing becomes possible, but also that the spreading algorithm starts from the most 'even' possible layout.

'Factor distance between systems': default 3
'Factor distance within system': default 1.5
'Factor distance staves grouped by [ bracket': default 1. This is meant to represent 'normal' spacing, so the others are relative to this.
(The English wording for these settings needs to be more clear as well, but I can't think at the moment of a really good way to do it)

'Max. staff distance' should have a higher default, at least 12
'Max. system distance' likewise, perhaps 24; but I find that setting this value high actually adds lots of space between the systems, and throws the proportions off - so I think it's being interpreted wrong.

**Comment by njvdberg** [comment]
With the current algorithm, just decreasing the minimum distances to 1.5 will not lead to a better result. The current algorithm starts with the existing layout and add for each category of staff distance (between systems, between [ grouped and the remaining gaps) the **same** space. The result can become strange:

Screenshot_20200905_144340

An alternative could be changing the algorithm. In stead of **adding** space to the existing distance, try to make the spacing between staves of the category as similar as possible first.

**Comment by oktophonie** [comment]
> An alternative could be changing the algorithm. In stead of **adding** space to the existing distance, try to make the spacing between staves of the category as similar as possible first.

That sounds like what I was trying to describe in the original specification :) i.e. start with close spacing, then increase the spaces towards as close as equal to possible (adjusting for the various factors).

**Comment by oktophonie** [comment]
I made a (crude) example to illustrate this better:
vertical-justification
The idea is that we set 'stave distance' to something small (1 or 1.5) so the staves are all close together, regardless of how irregular that makes them. That gives the 'Close spacing' version.

What this PR does (correct me if I'm wrong) is distribute the extra space equally (using the factors, which are irrelevant to this example), which gives the second version. As you can see, this preserves any irregularities from the initial close spacing, so it looks a bit lumpy.

But if we try to regularise the spaces as much as possible, you get the third version, so all the spaces tend towards whatever we determine the 'standard' is, and no extra space is added if we already have more than that.

**Comment by Tantacrul** [comment]
It is so great using this! I can't wait for it to be released and to tell users about it :) A few refining comments below.

--

There is one additional aspect to this that I think is worth doing: removing the possibility for staves to fall off the page (which is a really awful experience). There are two parts to this:

1. Prevent the staves from ever extending beyond the page margin at the bottom of the page and allow the instruments to continuously bunch together and overlap as more get added. I feel that this communicates the actual problem to users better than the staves disappearing off the page: namely that they need to alter their page size.

2. After a certain number of instruments have been added (totalling around 18 staves on the page), we show a dialog to users to recommend adjusting the page size to be bigger and provide an option to resize to A3 right there. I will happily design this flow and work on the exact options to put in front of users if you like?

My rationale here is that we have a lot of new / novice users, who sometimes need a bit of guidance. Experienced users will already know what page size they want and will not be affected adversely by something like this.

---

Simon and I also tried out different default settings in the 'Styles/Page' dialog. We felt the following worked best:
**Stave distance** 1.0
**Factor distance between systems:** 3.0
**Factor distance within system** 1.5
**Factor distance staves grouped by [ bracket** 1.0

Lastly (sorry for such a long comment!) - I agree with Simon that it would be lovely if we could achieve the spacing nuance he mentioned in his last comment. The spacing can look a little weird at times right now.

**Comment by IsaacWeiss** [comment]
I'd also suggest:

3. Reduce staff size continuously as more instruments get added
4. Show dialog recommending user to reduce staff size

I, for instance, would probably only have one size of paper on which to print, and would rather format the music to fit on that paper.

**Comment by Tantacrul** [comment]
> I'd also suggest:
> 
> 1. Reduce staff size continuously as more instruments get added
> 2. Show dialog recommending user to reduce staff size
> 
> I, for instance, would probably only have one size of paper on which to print, and would rather format the music to fit on that paper.

I largely agree, although I think we'd need to design that carefully. There are specific size standards for different page sizes. We'd definitely not want to auto-resize staves without explicitly asking first but, apart from that, it's a good point. We should generally provide the user with good sensible options to choose from and I think this is one situation where providing those options will be very useful to the user.

**Comment by njvdberg** [comment]
> That sounds like what I was trying to describe in the original specification :)

It seems I misread that part :smiley:, too much focused on improving the spacing of the *existing* layout. Back to the drawing board. Look at the bright side, it's only the algorithm. The hard part, actually changing the layout is still there.

>     1. Prevent the staves from ever extending beyond the page margin at the bottom of the page and allow the instruments to continuously bunch together and overlap as more get added. I feel that this communicates the actual problem to users better than the staves disappearing off the page: namely that they need to alter their page size.

I think the algorithm is doing exactly that but the lower distance limit prevent it. Removing these lower limit should do the job but I will check.

>     2. After a certain number of instruments have been added (totalling around 18 staves on the page), we show a dialog to users to recommend adjusting the page size to be bigger and provide an option to resize to A3 right there. I will happily design this flow and work on the exact options to put in front of users if you like?

This sound more an improvement for [PR #6479, Automatic score ordering and bracketing](https://github.com/musescore/MuseScore/pull/6479). That PR handles adding instruments. If I remember correctly, in the end, choosing a score order will not only set a certain score order but also some stylistic options. Paper size and staff size seems to me important setting for this.
So, I'm not sure this should be addressed here or in that other PR (or even in another PR if this kind of enhancements will appear in 4.0 rather than 3.6. This hold also for remarks 3 end 4 from @IsaacWeiss.

But as said, back to the drawing board. Watch this space :smiley:

**Comment by njvdberg** [comment]
New version of the spacing algorithm. Also update of style parameters as suggested by @Tantacrul  (and (@oktophonie)

File: libmscore/layout.h (hunk 1 above)
Comment (njvdberg) [reply libmscore/layout.h]: Has been removed.

**Comment by njvdberg** [comment]
There are 4 vtest violations:
 1. barline-1
 2. harmony-15
 3. system-9
 4. staff-1

All 4 are expected because on these 4 examples the spreading algorithm kicks in.

**Comment by chenlung** [comment]
I think it comes under [this](https://musescore.org/en/node/276106) issue.

I'm wondering about a tick box setting, but what do others think?

**Comment by njvdberg** [comment]
Implements algorithm based on the new spec.

File: libmscore/style.cpp [hunk 3]
```diff
@@ -77,6 +77,15 @@ static const StyleType styleTypes[] {
       { Sid::minSystemDistance,       "minSystemDistance",       Spatium(8.5)  },
       { Sid::maxSystemDistance,       "maxSystemDistance",       Spatium(15.0) },
 
+      { Sid::enableVerticalSpread,    "enableVerticalSpread",    false         },
+      { Sid::spreadSystem,            "spreadSystem",            2.5           },
+      { Sid::spreadSquareBracket,     "spreadSquareBracket",     1.0           },
+      { Sid::spreadCurleyBracket,     "spreadCurleyBracket",     1.0           },
```
Comment (Asmatzaile) [inline libmscore/style.cpp:83]: Curly?

Comment (njvdberg) [reply libmscore/style.cpp]: Good catch, corrected!

**Comment by njvdberg** [comment]
Now also contains the migration handlers for MuseScore 3.6.

**Pull request merged**